lib/core/include/experimental_plugin_framework.hpp

```diff
@@ -0,0 +1,461 @@
+        const std::string progress{"progress"};
+    }; // commands
+
+    namespace endpoints {
+        const std::string command{"command"};
+        const std::string operation{"operation"};
+    }; // endpoints
+
+    namespace constants {
+        const std::string code{"code"};
+        const std::string status{"status"};
+        const std::string plugin{"plugin"};
+        const std::string errors{"errors"};
+        const std::string command{"command"};
+        const std::string message{"message"};
+        const std::string progress{"progress"};
+    }; // constants
+
+    class base;
+
+    namespace {
+        void sleep()
+        {
+            std::this_thread::sleep_for(std::chrono::milliseconds(500));
+        }
+
+
+
+        auto invoke(rcComm_t& comm, const json& msg)
+        {
+            const uint32_t ADAPTOR_APN{120000};
+
+            auto str = msg.dump();
+
+            bytesBuf_t inp;
+            inp.buf = (void*)str.c_str();
+            inp.len = str.size();
+
+            bytesBuf_t* resp{};
+
+            auto err = procApiRequest(&comm, ADAPTOR_APN, (void*)&inp, NULL, (void**)&resp, NULL );
+
+            if (err < 0) {
+                THROW(err, "failed to perform the invocation");
+            }
+
+            return json::parse({(char*)resp->buf, (char*)resp->buf+resp->len});
+
+        } // invoke
+
+
+
+        template<typename T>
+        auto get(const std::string& n, const json& p)
+        {
+            if(!p.contains(n)) {
+                THROW(
+                    SYS_INVALID_INPUT_PARAM,
+                    boost::format("missing [%s] parameter")
+                    % n);
+            }
+
+            return p.at(n).get<T>();
+        } // get
+
+
+
+        template<typename T>
+        auto get(const std::string& n, const json& p, T d)
+        {
+            if(!p.contains(n)) {
+                return d;
+            }
+
+            return p.at(n).get<T>();
+        } // get
+
+
+
+        auto to_blackboard(const irods::future& f) -> json {
+            auto bb = json{{constants::status, states::complete}};
+
+            try {
+                auto xx = f.get();
+
+                if(xx.size() > 0) {
+                    auto arr = json::array();
+                    for(auto&& e : xx) {
+                        arr.push_back({{constants::code,    std::get<0>(e)},
+                                       {constants::message, std::get<1>(e)}});
+                    }
+
+                    bb[constants::errors] = arr;
+                    bb.update({{constants::status, states::failed}});
+                }
+            }
+            catch(...) {
+                return json{{constants::errors, {
+                            {constants::code, SYS_INTERNAL_ERR},
+                            {constants::message, "unknown error in to_blackboard"}}}};
+            }
+
+            return bb;
+
+        } // to_blackboard
+
+
+        #ifdef RODS_SERVER
```

Comment:
consider keeping macro indentation separate from runtime logic. this makes it easier to spot and informs the reader that they should handle the block of code carefully.